scalanet/src/io/iohk/scalanet/peergroup/udp/StaticUDPPeerGroup.scala

```diff
@@ -0,0 +1,444 @@
+  ): Task[Unit] =
+    for {
+      channels <- getChannels(remoteAddress)
+      _ <- Task.parTraverseUnordered(channels)(f).executeOn(scheduler)
+    } yield ()
+
+  /** Replicate the incoming message to the server channel and all client channels connected to the remote address. */
+  private def handleMessage(
+      remoteAddress: InetSocketAddress,
+      maybeMessage: Attempt[M]
+  ): Unit =
+    executeSync {
+      replicateToChannels(remoteAddress)(_.handleMessage(maybeMessage))
+    }
+
+  private def handleError(remoteAddress: InetSocketAddress, error: Throwable): Unit =
+    executeSync {
+      replicateToChannels(remoteAddress)(_.handleError(error))
+    }
+
+  // We cannot execute right on the channel's event loop because that's already
+  // occupied with handing the incoming message, so it would deadlock.
+  // The event loop will wait, but we have to delegate the execution elsewhere.
+  // The alternative would be to use the normal concurrent data structures
+  // that do internal locking without Tasks.
+  private def executeSync(task: Task[Unit]): Unit = {
+    task.runSyncUnsafe()
+  }
+
+  private def tryDecodeDatagram(datagram: DatagramPacket): Attempt[M] =
+    codec.decodeValue(BitVector(datagram.content.nioBuffer)) match {
+      case failure @ Attempt.Failure(err) =>
+        logger.debug(s"Message decoding failed due to ${err}", err)
+        failure
+      case success =>
+        success
+    }
+
+  private lazy val serverBinding =
+    new Bootstrap()
+      .group(workerGroup)
+      .channel(classOf[NioDatagramChannel])
+      .option[RecvByteBufAllocator](
+        ChannelOption.RCVBUF_ALLOCATOR,
+        new io.netty.channel.DefaultMaxBytesRecvByteBufAllocator()
+      )
+      .handler(new ChannelInitializer[NioDatagramChannel]() {
+        override def initChannel(nettyChannel: NioDatagramChannel): Unit = {
+          nettyChannel
+            .pipeline()
+            .addLast(new ChannelInboundHandlerAdapter() {
+              override def channelRead(ctx: ChannelHandlerContext, msg: Any): Unit = {
+                val datagram = msg.asInstanceOf[DatagramPacket]
+                val remoteAddress = datagram.sender
+                try {
+                  logger.debug(s"Server channel at $localAddress read message from $remoteAddress")
+                  handleMessage(remoteAddress, tryDecodeDatagram(datagram))
+                } catch {
+                  case NonFatal(ex) =>
+                    handleError(remoteAddress, ex)
+                } finally {
+                  datagram.content().release()
+                  ()
+                }
+              }
+
+              override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+                val channelId = ctx.channel().id()
+                val remoteAddress = ctx.channel.remoteAddress().asInstanceOf[InetSocketAddress]
+                cause match {
+                  case _: PortUnreachableException =>
```

Comment:
this case is probably not necessary as it was thrown only when using client bootstrap and doing connected version of udp.

Reply:
Okay great, I'll remove it, thanks!

Reply:
Removed, still calling `super`.